unified_planning/engines/mixins/simulator.py

```diff
@@ -0,0 +1,110 @@
+
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Iterator, List, Optional, Tuple, Union
+import unified_planning as up
+
+
+class Event:
+    '''This is an abstract class representing an event.'''
+
+    @property
+    def conditions(self) -> List['up.model.FNode']:
+        raise NotImplementedError
+
+    @property
+    def effects(self) -> List['up.model.Effect']:
+        raise NotImplementedError
+
+    @property
+    def simulated_effect(self) -> Optional['up.model.SimulatedEffect']:
+        raise NotImplementedError
+
+
+class Simulator:
+    '''
+    Simulator abstract class.
+    This class defines the interface that a simulator must implement.
+
+    Important NOTE: The AbstractProblem instance is given at the constructor.
+    '''
+
+    def __init__(self, problem: 'up.model.AbstractProblem', **params) -> None:
+        '''
+        Takes an instance of a problem and eventually some parameters, that represent
+        some specific settings of the Simulator.
+        :param problem: the problem that defines the domain in which the simulation exists.
+        :param **params: a dict[str, str] containing the Simulator settings.
+        '''
+        raise NotImplementedError
```

Comment:
```suggestion
        self._problem = problem
```

and update the SequentialSimulator accordingly.